### DISCUSSION TITLE
UI / Button

### DISCUSSION THREAD

**Discussion by lennoximus** [dcrtit/kit#6] [Ideas]
# Слоты

![изображение](https://user-images.githubusercontent.com/14184055/152032593-c5882029-f28e-449f-97ef-982e4af6ec67.png)

Слот | Описание | Назначение
-- | -- | --
prefloat | Слот под элементы, обтекаемые button, но стилистически находящийся вне его, спозиционирован перед кнопкой | Инпуты, иконки, таймеры
prepend-icon | Слот под иконки, находящийся после слота с текстом кнопки, прижат к левому краю, стилистически внутри кнопки | Иконки, буквенные шрифтовые инфографики
text | Слот под текст, либо иконку, когда речь идет о social-buttons и др. нужд | Текст, иконка
text-float-icon | Слот под иконки, находящиейся в контексте текста кнопки, стоящий после текстовой части (пока сомнительна его необходимость, изначально хотелось управлять отступами автоматически, по гайдлайнам, чтобы этого не пришлось делать в последствии юзеру) | Иконки, буквенные шрифтовые инфографики
append-icon | Слот под иконки, находящийся перед слотом с текстом кнопки, прижат к правому краю, стилистически внутри кнопки | Иконки, буквенные шрифтовые инфографики
postfloat | Слот под элементы, обтекаемые button, но стилистически находящийся вне его, спозиционирован после кнопки | Инпуты, иконки, таймеры

# Пропсы

```
*custom string* подразумевает, что в сборке так же будут участвовать кастомные пары 'bg-color - font-color', по аналогии vuetify
https://vuetifyjs.com/en/features/theme/#light-and-dark
```

Проп | Тип | Описание
-- | -- | --
color | 'success' \| 'primary' \| 'warn' \| 'error' \| 'grey' \| 'light' \| 'dark' \| 'transparent' \| *custom string* | Цвет кнопки, все значения кроме *custom string* управляют так же цветом шрифта для контраста
size | 'xlarge' \| 'large' \| 'medium' \| 'default' \| 'small' \| 'xsmall' \| *custom string* | Размер кнопки, так же управляет размером шрифта, кроме *custom string*
block | Boolean | Управляет размером кнопки - по ширине блока
loading | Boolean | Управляет состоянием loading
disabled | Boolean | Управляет состоянием disabled (под сомнением)

**Comment by DiceMasters** [comment]
Касаемо пропсов управления стилями кнопки, предложение следующее, сделать под них отдельные классы типа
```
interface ComponentTemplateColor {
  bg: string // hex | rgb | rgba
  font: string // hex | rgb | rgba
}
```
```
interface ComponentTemplateSize {
  height: number
  padding: string // css paddings
  fontsize: number
}
```
Эти классы, либо один большой класс фабрику можно будет расширять для расширения коллекций пропсов, не совсем академически верно выразился, но надеюсь суть поняли

**Comment by DiceMasters** [comment]
А не забейте, бесполезно получается как то

**Comment by lennoximus** [comment]
Я думаю, что `color` и `size` надо продумать, как будем кастомизировать. Либо какой-то конфиг заводить, где будем посасывать возможные значения -> в стилях это тоже как-то обрабатывать, либо вообще отказаться от этой затеи, так как она кмк не особо масштабируется с учетом headless-подхода.

**Reply by lennoximus** [reply]
Обсудим на общем созвоне.
